### PR TITLE
NIAD-1087: translating stubbed codeable concept to cd

### PR DESCRIPTION
**Pull request by ElliotB-96**
- Created mapper
- Created unit tests for mapper
- Added mapper usage to other mappers that have codeable concepts to transform
- Updated unit tests for other mappers to validate call to new mapper

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureMapper.java [hunk 1]
```diff
@@ -187,4 +189,11 @@ private String buildNotePertinentInformation(AllergyIntolerance allergyIntoleran
         }
         return notes;
     }
+
+    private String buildCode(AllergyIntolerance allergyIntolerance) {
+        if (allergyIntolerance.hasCode()) {
+            return codeableConceptCdMapper.mapCodeableConceptToCd(allergyIntolerance.getCode());
+        }
+        return StringUtils.EMPTY;
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureMapper.java:197]: Where properties are required (Card 1..1) the adaptor should throw an exception if not present.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ConditionLinkSetMapper.java [hunk 2]
```diff
@@ -167,4 +170,11 @@ private boolean filterOutListResourceType(Reference reference) {
     private boolean checkIfReferenceIsObservation(Reference reference) {
         return reference.getReferenceElement().getResourceType().equals(ResourceType.Observation.name());
     }
+
+    private Optional<String> buildCode(Condition condition) {
+        if (condition.hasCode()) {
+            return Optional.of(codeableConceptCdMapper.mapCodeableConceptToCd(condition.getCode()));
+        }
+        return Optional.empty();
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ConditionLinkSetMapper.java:178]: As above, 'code' is required in Condition so an error should be thrown

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/DiaryPlanStatementMapper.java [hunk 3]
```diff
@@ -154,6 +156,13 @@ private String formatReason(String value) {
         return Optional.empty();
     }
 
+    private Optional<String> buildCode(ProcedureRequest procedureRequest) {
+        if (procedureRequest.hasCode()) {
+            return Optional.of(codeableConceptCdMapper.mapCodeableConceptToCd(procedureRequest.getCode()));
+        }
+        return Optional.empty();
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/DiaryPlanStatementMapper.java:163]: code is required, should throw

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java [hunk 4]
```diff
@@ -30,9 +32,17 @@ public String mapEncounterToEhrComposition(Encounter encounter) {
...
             .availabilityTime(StatementTimeMappingUtils.prepareAvailabilityTimeForEncounter(encounter))
             .status(COMPLETE_CODE)
+            .type(buildType(encounter))
             .components(components);
 
         return TemplateUtils.fillTemplate(ENCOUNTER_STATEMENT_TO_EHR_COMPOSITION_TEMPLATE,
             encounterStatementTemplateParameters.build());
     }
+
+    private String buildType(Encounter encounter) {
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java:42]: Encounter.type is a special case handled by NIAD-1113 and should have been omitted by 1087. I'll update the story and these changed to EncounterMapper can be rolled back.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 5]
```diff
@@ -223,4 +228,27 @@ private String extractVaccinationProtocolString(Immunization.ImmunizationVaccina
...
         return vaccinationProtocol + VACCINATION_TARGET_DISEASE + targetDiseases;
     }
+
+    private String buildCode(Immunization immunization) {
+        var vaccinationProcedure = ExtensionMappingUtils.filterExtensionByUrl(immunization, VACCINATION_PROCEDURE_URL);
+        if (vaccinationProcedure.isPresent()) {
+            CodeableConcept codeableConcept = (CodeableConcept) vaccinationProcedure.get().getValue();
+            return codeableConceptCdMapper.mapCodeableConceptToCd(codeableConcept);
+        }
+        return StringUtils.EMPTY;
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:238]: The vaccinationProcedure extension is required. Should throw error is missing.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 6]
```diff
@@ -223,4 +228,27 @@ private String extractVaccinationProtocolString(Immunization.ImmunizationVaccina
...
+        if (immunization.hasVaccineCode() && vaccineCodeNotUNK(immunization.getVaccineCode())) {
+            var code = CodeableConceptMappingUtils.extractTextOrCoding(immunization.getVaccineCode());
+            if (code.isPresent()) {
+                return String.format(VACCINATION_CODE, code.get());
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private boolean vaccineCodeNotUNK(CodeableConcept codeableConcept) {
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:251]: The vaccinationProcedure extension is required. Should throw error is missing. Return empty only for UNK code.

File: service/src/main/resources/templates/ehr_allergy_structure_template.mustache [hunk 7]
```diff
@@ -28,6 +28,7 @@
                             <text>{{pertinentInformation}}</text>
                         </pertinentAnnotation>
                     </pertinentInformation>
+                    {{{code}}}
```
Comment (z-uddin) [inline service/src/main/resources/templates/ehr_allergy_structure_template.mustache:31]: Should there also be a code for the ehrComposition level?

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ConditionLinkSetMapper.java [hunk 8]
```diff
@@ -167,4 +171,11 @@ private boolean filterOutListResourceType(Reference reference) {
     private boolean checkIfReferenceIsObservation(Reference reference) {
         return reference.getReferenceElement().getResourceType().equals(ResourceType.Observation.name());
     }
+
+    private Optional<String> buildCode(Condition condition) {
+        if (condition.hasCode()) {
+            return Optional.of(codeableConceptCdMapper.mapCodeableConceptToCd(condition.getCode()));
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ConditionLinkSetMapper.java:177]: I don't think mapCodeableConceptToCd should ever return null. Do we need to use Optional here?

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/DiaryPlanStatementMapper.java [hunk 9]
```diff
@@ -154,6 +156,13 @@ private String formatReason(String value) {
         return Optional.empty();
     }
 
+    private Optional<String> buildCode(ProcedureRequest procedureRequest) {
+        if (procedureRequest.hasCode()) {
+            return Optional.of(codeableConceptCdMapper.mapCodeableConceptToCd(procedureRequest.getCode()));
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/DiaryPlanStatementMapper.java:161]: as above re optional

File: service/src/main/resources/templates/ehr_request_statement_template.mustache [hunk 10]
```diff
@@ -7,5 +7,8 @@
         {{{availabilityTime}}}
         {{{defaultReasonCode}}}
         <text>{{description}}</text>
+        {{#code}}
```
Comment (mattd-kainos) [inline service/src/main/resources/templates/ehr_request_statement_template.mustache:10]: Do we need the code section here?

File: service/src/main/resources/templates/immunization_observation_statement_template.mustache [hunk 11]
```diff
@@ -6,11 +6,16 @@
...
-</pertinentInformation>{{/pertinentInformation}}
+{{#pertinentInformation}}
+    <pertinentInformation typeCode="PERT">
+        <sequenceNumber value="+1"/>
+        <pertinentAnnotation classCode="OBS" moodCode="EVN">
+            <text>{{pertinentInformation}}</text>
+        </pertinentAnnotation>
+<   /pertinentInformation>
+{{/pertinentInformation}}
+        {{#code}}
```
Comment (mattd-kainos) [inline service/src/main/resources/templates/immunization_observation_statement_template.mustache:17]: is the code section required here?

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureMapperTest.java [hunk 12]
```diff
@@ -82,7 +92,7 @@ public void When_MappingAllergyIntoleranceJson_Expect_AllergyStructureXmlOutput(
 
         String outputMessage = allergyStructureMapper.mapAllergyIntoleranceToAllergyStructure(parsedAllergyIntolerance);
 
-        assertThat(outputMessage).isEqualTo(expectedOutputMessage);
+        assertThat(outputMessage).isEqualToIgnoringWhitespace(expectedOutputMessage);
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureMapperTest.java:95]: We can't use `isEqualToIgnoringWhitespace` in our mapper tests. There are places where it is signification e.g. notes / free text sections.

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureMapperTest.java [hunk 13]
```diff
@@ -16,13 +18,17 @@
...
+import org.mockito.quality.Strictness;
 
 import uk.nhs.adaptors.gp2gp.common.service.FhirParseService;
 import uk.nhs.adaptors.gp2gp.common.service.RandomIdGeneratorService;
 import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
+import uk.nhs.adaptors.gp2gp.utils.CodeableConceptMapperMockUtil;
 import uk.nhs.adaptors.gp2gp.utils.ResourceTestFileUtils;
 
 @ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureMapperTest.java:31]: Instead of using lenient strictness only mock behaviour for the test cases that require it

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ConditionLinkSetMapperTest.java [hunk 14]
```diff
@@ -23,9 +25,12 @@
 import org.junit.jupiter.params.provider.MethodSource;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
 import org.mockito.stubbing.Answer;
 
 @ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ConditionLinkSetMapperTest.java:33]: As above, if you need to use lenient strictness, that mean the test is mocking behaviour that is unused. This usually indicates that the test is written incorrectly.

File: service/src/test/resources/ehr/mapper/immunization/expected-output-observation-statement-no-information.xml [hunk 15]
```diff
@@ -6,5 +6,14 @@
             <center value="20100630055900"/>
         </effectiveTime>
         <availabilityTime value="20100630054806" />
+        <pertinentInformation typeCode="PERT">
+            <sequenceNumber value="+1"/>
+            <pertinentAnnotation classCode="OBS" moodCode="EVN">
+                <text>Substance: unknown</text>
```
Comment (mattd-kainos) [inline service/src/test/resources/ehr/mapper/immunization/expected-output-observation-statement-no-information.xml:12]: Can you confirm that "Substance: unknown" is the correct output in this case?

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java [hunk 16]
```diff
@@ -17,13 +19,17 @@
...
+import org.mockito.quality.Strictness;
 
 import uk.nhs.adaptors.gp2gp.common.service.FhirParseService;
 import uk.nhs.adaptors.gp2gp.common.service.RandomIdGeneratorService;
 import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
+import uk.nhs.adaptors.gp2gp.utils.CodeableConceptMapperMockUtil;
 import uk.nhs.adaptors.gp2gp.utils.ResourceTestFileUtils;
 
 @ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java:32]: as above about lenient strictness

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementMapperTest.java [hunk 17]
```diff
@@ -18,13 +20,17 @@
...
+import org.mockito.quality.Strictness;
 
 import uk.nhs.adaptors.gp2gp.common.service.FhirParseService;
 import uk.nhs.adaptors.gp2gp.common.service.RandomIdGeneratorService;
 import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
+import uk.nhs.adaptors.gp2gp.utils.CodeableConceptMapperMockUtil;
 import uk.nhs.adaptors.gp2gp.utils.ResourceTestFileUtils;
 
 @ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementMapperTest.java:33]: as above about lenient strictness

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementMapperTest.java [hunk 18]
```diff
@@ -105,7 +115,7 @@ public void When_MappingObservationJson_Expect_NarrativeStatementXmlOutput(Strin
 
         String outputMessage = requestStatementMapper.mapReferralRequestToRequestStatement(parsedReferralRequest, false);
 
-        assertThat(outputMessage).isEqualTo(expectedOutputMessage);
+        assertThat(outputMessage).isEqualToIgnoringWhitespace(expectedOutputMessage);
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementMapperTest.java:118]: don't ignore whitespace

File: service/src/test/java/uk/nhs/adaptors/gp2gp/utils/CodeableConceptMapperMockUtil.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package uk.nhs.adaptors.gp2gp.utils;
+
+public class CodeableConceptMapperMockUtil {
+    public static final String NULL_FLAVOR_CODE = "<code nullFlavor=\"UNK\">\n"
+        + "    <originalText>4</originalText>\n"
+        + "</code>";
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/utils/CodeableConceptMapperMockUtil.java:6]: Putting on a single line might help with the formatting headaches, at least for the mock ones.

```suggestion
    public static final String NULL_FLAVOR_CODE = "<code nullFlavor=\"UNK\"><originalText>4</originalText></code>";
```

Comment (mattd-kainos) [reply service/src/test/java/uk/nhs/adaptors/gp2gp/utils/CodeableConceptMapperMockUtil.java:6]: Optional improvement: 4 => "Mocked code" so it is more obvious where this value comes from

File: service/src/test/resources/ehr/mapper/immunization/expected-output-observation-statement-no-information.xml (hunk 15 above)
Comment (ElliotB-96) [reply service/src/test/resources/ehr/mapper/immunization/expected-output-observation-statement-no-information.xml:12]: VaccineCode is a required field

File: service/src/test/resources/ehr/mapper/condition/condition_1_resource_1_list.json [hunk 20]
```diff
@@ -82,5 +61,8 @@
         {
             "text": "Burning feet description"
         }
-    ]
+    ],
```
Comment (mattd-kainos) [inline service/src/test/resources/ehr/mapper/condition/condition_1_resource_1_list.json:64]: Were these change required for all of the condition JSON fixture or could they be left as they were?

File: service/src/test/resources/ehr/mapper/encountercomponents/expected-components-1.xml [hunk 21]
```diff
@@ -50,38 +51,46 @@
...
     <effectiveTime>
         <center value="20100630055900"/>
     </effectiveTime>
-    <availabilityTime value="20100630054806"/>
+    <availabilityTime value="20100630054806" />
     <pertinentInformation typeCode="PERT">
         <sequenceNumber value="+1"/>
         <pertinentAnnotation classCode="OBS" moodCode="EVN">
-            <text>immunization note</text>
+            <text>immunization note Substance: unknown</text>
```
Comment (mattd-kainos) [inline service/src/test/resources/ehr/mapper/encountercomponents/expected-components-1.xml:70]: Should this text actually be prepended to the note? Verify this one.

File: service/src/test/resources/ehr/mapper/encountercomponents/expected-components-1.xml [hunk 22]
```diff
@@ -1,29 +1,30 @@
-<component typeCode="COMP">
+<component typeCode="COMP" >
```
Comment (mattd-kainos) [inline service/src/test/resources/ehr/mapper/encountercomponents/expected-components-1.xml:1]: Were all of these trailing spaces added by running the IDE formatter or similar? If so please revert, it makes it difficult to see the relevant changes for this feature.

File: service/src/main/resources/templates/ehr_request_statement_template.mustache (hunk 10 above)
Comment (ElliotB-96) [reply service/src/main/resources/templates/ehr_request_statement_template.mustache:10]: ReasonCode has 0..* cardinality and if there isn't a block here it leaves a blank line that breaks the tests

File: service/src/test/resources/ehr/mapper/condition/condition_1_resource_1_list.json (hunk 20 above)
Comment (ElliotB-96) [reply service/src/test/resources/ehr/mapper/condition/condition_1_resource_1_list.json:64]: I have reverted them

File: service/src/test/resources/ehr/mapper/encountercomponents/expected-components-1.xml (hunk 21 above)
Comment (ElliotB-96) [reply service/src/test/resources/ehr/mapper/encountercomponents/expected-components-1.xml:70]: Would be worth mentioning that this is also gone now as UNK is not mapped.

Comment (ElliotB-96) [reply service/src/test/resources/ehr/mapper/encountercomponents/expected-components-1.xml]: Verified by David, render as text element unless unknown.

File: service/src/main/resources/templates/ehr_allergy_structure_template.mustache (hunk 7 above)
Comment (z-uddin) [reply service/src/main/resources/templates/ehr_allergy_structure_template.mustache:31]: Confirmed with @dmcavenue1 that code does not need to be added at ehrComposition level for allergy structure.

**Review by mattd-kainos — APPROVED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/CodeableConceptCdMapper.java [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+            var displayText = findDisplayText(mainCode.get());
+
+            code.ifPresent(builder::mainCode);
+            displayText.ifPresent(builder::mainDisplayName);
+        }
+
+        return TemplateUtils.fillTemplate(CODEABLE_CONCEPT_CD_TEMPLATE, builder.build());
+    }
+
+    public String mapEncounterTypeNoSnomed(Encounter encounter) {
```
Comment (msolecka) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/CodeableConceptCdMapper.java:54]: id this method not used?

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 24]
```diff
@@ -223,4 +228,29 @@ private String extractVaccinationProtocolString(Immunization.ImmunizationVaccina
...
+                return String.format(VACCINATION_CODE, code.get());
+            }
+        } else if (immunization.hasVaccineCode() && !vaccineCodeNotUNK(immunization.getVaccineCode())) {
+            return StringUtils.EMPTY;
+        }
+        throw new EhrMapperException("Immunization vaccine code not present");
+    }
+
+    private boolean vaccineCodeNotUNK(CodeableConcept codeableConcept) {
+        return  (codeableConcept.getCodingFirstRep().hasCode() && !codeableConcept.getCodingFirstRep().getCode().equals("UNK"));
```
Comment (msolecka) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:254]: why is there double negation? wouldn't be enough to check id isVaccineCodeUNK?

File: service/src/test/resources/ehr/mapper/codeableconcept/codeable_concept_with_text.json [hunk 25]
```diff
@@ -0,0 +1,35 @@
+{
+    "resourceType" : "Observation",
+    "code": {
+        "coding": [
+            {
+                "system": "http://blah.info/blahsomecodesystem",
+                "code": "X%^^%",
+                "display": "Endometriosis of uterus",
```
Comment (msolecka) [inline service/src/test/resources/ehr/mapper/codeableconcept/codeable_concept_with_text.json:8]: there could be different display to verify if the taken display is really snomed code

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java (hunk 24 above)
Comment (Broom94) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:254]: I think this will through npe

File: service/src/test/resources/ehr/request/fhir/output/ExpectedEhrExtractResponseFromJson.xml [hunk 26]
```diff
@@ -378,14 +396,18 @@
...
                                 <sequenceNumber value="+1" />
                                 <pertinentAnnotation classCode="OBS" moodCode="EVN">
                                     <text>This is the BP reading in a consultation with a note</text>
                                 </pertinentAnnotation>
                             </pertinentInformation>
+
+                            <code nullFlavor="UNK"><originalText>Mocked code</originalText></code>
                         </ObservationStatement>
                     </component>
+
```
Comment (Broom94) [inline service/src/test/resources/ehr/request/fhir/output/ExpectedEhrExtractResponseFromJson.xml:410]: This one and a few xmls above have some random whitespaces, as discusses please investigate

**Review by msolecka — DISMISSED** [review]

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EhrExtractMapperTest.java [hunk 27]
```diff
@@ -100,7 +108,7 @@ public void When_MappingProperJsonRequestBody_Expect_ProperXmlOutput() throws IO
             getGpcStructuredTaskDefinition,
             bundle);
         String output = ehrExtractMapper.mapEhrExtractToXml(ehrExtractTemplateParameters);
-
+        System.out.println(output);
```
Comment (mattd-kainos) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EhrExtractMapperTest.java:111]: remove this

**Review by mattd-kainos — APPROVED** [review]

**Review by msolecka — APPROVED** [review]
